### PR TITLE
Virtio-forwarder

### PR DESCRIPTION
**Pull request by noaezra**

### CODE REVIEW

File: lib/netdev-rte-offloads.c [hunk 1]
```diff
@@ -1863,16 +1865,14 @@ static struct rte_eth_conf netdev_rte_offloads_port_conf = {
 struct qpair_stats {
     uint64_t packets;
     uint64_t bytes;
-    uint64_t cycles;
```
Comment (elibritstein) [inline lib/netdev-rte-offloads.c]: please split non related changes to different commits

File: lib/netdev-dpdk.c [hunk 2]
```diff
@@ -1388,6 +1392,13 @@ netdev_dpdk_destruct(struct netdev *netdev)
 
     ovs_mutex_lock(&dpdk_mutex);
 
+    if (unlikely(dev->set_fwd != NULL)) {
+        dev->set_fwd = NULL;
+        dev->del_relay(dev->relay_id);
+        dev->del_relay = NULL;
+        dev->relay_id = 0xFFFF; //NOA: change to define
```
[the remark this answers is not in the document]
Comment (noaezra) [reply lib/netdev-dpdk.c]: because 0 is a valid id

[the remark this answers is not in the document]
Comment (elibritstein) [reply lib/netdev-dpdk.c]: i understand. i suggested to make it invalid.
don't ever allocate it in netdev_rte_offloads_add_relay

[the remark this answers is not in the document]
Comment (elibritstein) [reply lib/netdev-dpdk.c]: or better:
- change "relay_id" to a pointer to your relay struct
- remove static array relays
- when allocating, do rte_zmalloc for a single relay struct, and keep its pointer in dev
- when freeing, call rte_free
- NULL will be your indication for invalid

you avoid allocating an array that most probably won't be utilized
you avoid accessing [relay_id], but direct pointer

File: lib/netdev-dpdk.c [hunk 3]
```diff
@@ -4249,6 +4264,50 @@ netdev_dpdk_rte_flow_create(struct netdev *netdev,
...
+
+void netdev_dpdk_hw_forwarder_register(const char *name, int id,
+        void (*fwd_fp)(int queue_id, int relay_id),
+        void (*del_fp)(int relay_id))
+{
+    struct netdev *netdev = NULL;
+    struct netdev_dpdk *dev = NULL;
+
+    netdev = netdev_from_name(name);
+    if (!is_dpdk_class(netdev->netdev_class)) {
```
Comment (rbaryanai) [inline lib/netdev-dpdk.c]: you need to release the ref here.
nedev_from_name inc the ref count.

File: lib/netdev-dpdk.c [hunk 4]
```diff
@@ -4249,6 +4264,50 @@ netdev_dpdk_rte_flow_create(struct netdev *netdev,
...
+}
+
+void netdev_dpdk_hw_forwarder_register(const char *name, int id,
+        void (*fwd_fp)(int queue_id, int relay_id),
+        void (*del_fp)(int relay_id))
+{
+    struct netdev *netdev = NULL;
+    struct netdev_dpdk *dev = NULL;
+
+    netdev = netdev_from_name(name);
```
Comment (rbaryanai) [inline lib/netdev-dpdk.c]: better check also the netdev is not NULL, because if user gave the wrong name
it will crash on next line

File: lib/netdev-dpdk.c [hunk 5]
```diff
@@ -4249,6 +4264,50 @@ netdev_dpdk_rte_flow_create(struct netdev *netdev,
...
+    if (!is_dpdk_class(netdev->netdev_class)) {
+        VLOG_ERR("netdev is not dpdk class\n");
+        return;
+    }
+    dev = netdev_dpdk_cast(netdev);
+    if (dev == NULL) {
+        VLOG_ERR("netdev_dpdk_hw_forwarder_register failed\n");
+        //unlock
```
[the remark this answers is not in the document]
Comment (rbaryanai) [reply lib/netdev-dpdk.c]: you don't need to check after cast if dev is null

File: lib/netdev-rte-offloads.c [hunk 6]
```diff
@@ -1825,3 +1841,879 @@ netdev_rte_offload_preprocess(struct dp_packet *packet, uint32_t mark)
...
+    rte_eth_dev_stop(port_id);
+    rte_eth_dev_callback_unregister(port_id, RTE_ETH_EVENT_QUEUE_STATE,
+            netdev_rte_offloads_queue_state_cb_fn, &relays[relay_id]);
+    rte_eth_dev_callback_unregister(port_id, RTE_ETH_EVENT_INTR_LSC,
+            netdev_rte_offloads_link_status_cb_fn, &relays[relay_id]);
+    rte_eth_dev_close(port_id);
+
+    if (relays[relay_id].flow_params.flow != NULL ) {
+        struct rte_flow_error error;
+        rte_flow_destroy(port_id, relays[relay_id].flow_params.flow, &error);
```
Comment (rbaryanai) [inline lib/netdev-rte-offloads.c]: better set flow = NULL after to destroy it (even that currently you don't touch it on the clean).

File: lib/netdev-rte-offloads.c [hunk 7]
```diff
@@ -1825,3 +1841,879 @@ netdev_rte_offload_preprocess(struct dp_packet *packet, uint32_t mark)
...
+    sprintf(vhost_name, "net_vhost%d", relay_id);
+    rte_eal_hotplug_remove("vdev", vhost_name);
+    rte_eal_hotplug_remove("pci", relays[relay_id].vf_pci);
+
+    netdev_rte_offloads_clear_relay(&relays[relay_id]);
+    return 0;
+}
+
+static void
+netdev_rte_offloads_show_statistics_vm(struct ds *reply)
```
Comment (rbaryanai) [inline lib/netdev-rte-offloads.c]: do you protect some how that while reading stats no one remove the pair?
maybe you need a lock here.

File: lib/netdev-rte-offloads.c [hunk 8]
```diff
@@ -1825,3 +1841,879 @@ netdev_rte_offload_preprocess(struct dp_packet *packet, uint32_t mark)
...
+                    count_stats[TX_BYTES] += qp->qp_stats.bytes;
+                } else{
+                    count_stats[RX_PKTS] += qp->qp_stats.packets;
+                    count_stats[RX_BYTES] += qp->qp_stats.bytes;
+                }
+            }
+
+            ds_put_format(reply,"\tRX-packets: %lu  RX-bytes: %lu\n",
+                    count_stats[RX_PKTS], count_stats[RX_BYTES]);
+            ds_put_format(reply,"\tTX-packets: %lu  TX-bytes: %lu\n",
```
Comment (rbaryanai) [inline lib/netdev-rte-offloads.c]: does "%lu" works here, 64 for should "%llu" I think

File: lib/netdev-rte-offloads.c [hunk 9]
```diff
@@ -1825,3 +1841,879 @@ netdev_rte_offload_preprocess(struct dp_packet *packet, uint32_t mark)
...
+    for (relay_id = 0; relay_id < NETDEV_RTE_OFFLOADS_MAX_RELAYS; ++relay_id) {
+        if (rte_atomic16_read(&relays[relay_id].valid) == 1) {
+            ds_put_format(reply, "relay id: %u, VM:  %s, VF:  %s\n",
+                    relay_id, relays[relay_id].vm_socket,
+                    relays[relay_id].vf_pci);
+
+            memset(count_stats, 0, sizeof count_stats);
+            for (qp_index = 0; qp_index < NETDEV_RTE_OFFLOADS_MAX_QPAIRS; qp_index++) {
+                qp = &relays[relay_id].qpair[qp_index];
+                if (qp_index&1) {
```
Comment (rbaryanai) [inline lib/netdev-rte-offloads.c]: qp_index & 1

File: lib/netdev-rte-offloads.c [hunk 10]
```diff
@@ -1825,3 +1841,879 @@ netdev_rte_offload_preprocess(struct dp_packet *packet, uint32_t mark)
...
+                             0, 0, netdev_rte_offloads_cmd,
+                             NULL);
+    initialized = 1;
+}
+
+void
+netdev_rte_offloads_hw_pr_fwd(int queue_id, int relay_id, int *fwd_rx)
+{
+    uint32_t q;
+    for (q = 0; q < (relays[relay_id].num_queues * 2); ++q) {
```
Comment (rbaryanai) [inline lib/netdev-rte-offloads.c]: why you can't have direct access here?

File: lib/netdev-rte-offloads.c [hunk 11]
```diff
@@ -1825,3 +1841,879 @@ netdev_rte_offload_preprocess(struct dp_packet *packet, uint32_t mark)
...
+    uint32_t i;
+
+    for (i = 0; i < argc; i++) {
+        ds_put_cstr(&reply, argv[i]);
+        ds_put_cstr(&reply, " ");
+    }
+    ds_put_cstr(&reply, "\n");
+
+    if (strcmp(argv[0], "hw-offload/set-forwarder") == 0) {
+        max_queues = atoi(argv[4]);
```
Comment (rbaryanai) [inline lib/netdev-rte-offloads.c]: how do you know you have argv[4]?

File: lib/netdev-rte-offloads.c [hunk 12]
```diff
@@ -1825,3 +1841,879 @@ netdev_rte_offload_preprocess(struct dp_packet *packet, uint32_t mark)
...
+            break;
+        }
+    }
+    if (!found) {
+        VLOG_ERR("Could not find relay for %s", vhost_path);
+        return -1;
+    }
+
+    /* if the PR still exist, remove the relay from the pr */
+    if (!remove_dev) {
```
Comment (rbaryanai) [inline lib/netdev-rte-offloads.c]: confusing. you call register to unregister.

File: lib/netdev-rte-offloads.c (hunk 11 above)
Comment (noaezra) [reply lib/netdev-rte-offloads.c]: it's a must for set_forwarder.
if the user won't set 5 args hw will get an error:
"hw-offload/set-forwarder" command requires at least 5 arguments

File: lib/netdev-rte-offloads.c (hunk 10 above)
Comment (noaezra) [reply lib/netdev-rte-offloads.c]: because I don't know which qp have this PR queue. I need to go over the qps in order to find the who is associated with this PR queue (can be more than one)

File: lib/netdev-rte-offloads.c [hunk 13]
```diff
@@ -1825,3 +1839,822 @@ netdev_rte_offload_preprocess(struct dp_packet *packet, uint32_t mark)
...
+
+    /*receive burst:*/
+    burst_success = rte_eth_rx_burst(qpair->pid_rx, queue_id,
+            qpair->pkts+qpair->mbuf_tail, NETDEV_MAX_BURST);
+    qpair->mbuf_tail += burst_success;
+    diff += burst_success;
+
+    /*send:*/
+send:
+    burst_success = rte_eth_tx_burst(qpair->pid_tx, queue_id,
```
[the remark this answers is not in the document]
Comment (noaezra) [reply lib/netdev-rte-offloads.c]: after talking to Roni we decided that rte_eth_tx_burst is more suitable for this case

**Comment by noaezra** [comment]
PR has been updated.

File: lib/netdev-dpdk.c [hunk 14]
```diff
@@ -4652,6 +4667,49 @@ netdev_dpdk_rte_flow_create(struct netdev *netdev,
...
+        VLOG_ERR("netdev was not found for %s\n", name);
+        return;
+    }
+    if (!is_dpdk_class(netdev->netdev_class)) {
+        VLOG_ERR("netdev is not dpdk class\n");
+        goto out;
+    }
+    dev = netdev_dpdk_cast(netdev);
+
+    if (fwd_fp != NULL) {
```
Comment (elibritstein) [inline lib/netdev-dpdk.c]: check == NULL (error) and proceed otherwise

File: lib/netdev-rte-offloads.c [hunk 15]
```diff
@@ -18,6 +18,15 @@
...
 #include <rte_flow.h>
+#include <rte_eth_vhost.h>
+#include <rte_ethdev.h>
+#include <rte_errno.h>
+#include <rte_mbuf.h>
+#include <rte_malloc.h>
+#include <rte_atomic.h>
+#include <rte_common.h>
+#include <rte_ether.h>
+#include <rte_byteorder.h>
```
Comment (elibritstein) [inline lib/netdev-rte-offloads.c]: i'm not sure we need all of those.
anyway, we can put them together in the same hunk as all the forwarder (bottom of the file)

File: lib/netdev-dpdk.c [hunk 16]
```diff
@@ -4652,6 +4667,49 @@ netdev_dpdk_rte_flow_create(struct netdev *netdev,
     return flow;
 }
 
+struct rte_mempool *netdev_dpdk_hw_forwarder_get_mempool(const char *name)
+{
+    struct netdev *netdev = NULL;
+    struct netdev_dpdk *dev = NULL;
```
Comment (elibritstein) [inline lib/netdev-dpdk.c]: rev xmas tree (all over)

File: lib/netdev-rte-offloads.c [hunk 17]
```diff
@@ -1778,3 +1788,838 @@ netdev_rte_offload_preprocess(struct dp_packet *packet, uint32_t mark)
...
+        VLOG_ERR("No port id found for vm %s\n", vhost_name);
+        ret = -1;
+        goto out;
+    }
+
+    /* create vf:*/
+    ret = rte_eal_hotplug_add("pci", pci, NULL);
+    if (ret) {
+        VLOG_ERR("rte_eal_hotplug_add pci failed\n");
+        rte_eal_hotplug_remove("vdev", vhost_name);
```
Comment (elibritstein) [inline lib/netdev-rte-offloads.c]: error handling at the rollback labels. can have few for different stages of rollback

File: lib/netdev-rte-offloads.c [hunk 18]
```diff
@@ -1778,3 +1788,838 @@ netdev_rte_offload_preprocess(struct dp_packet *packet, uint32_t mark)
...
+        relay->qpair[q].qp_stats.packets = 0;
+    }
+
+    memset(relay->vm_socket, 0, sizeof relay->vm_socket);
+    memset(relay->vf_pci, 0, sizeof relay->vf_pci);
+    relay->port_id_vm = 0;
+    relay->port_id_vf = 0;
+    relay->num_queues = 0;
+    relay->mempool = NULL;
+    rte_free(relay->pr_netdev);
```
Comment (elibritstein) [inline lib/netdev-rte-offloads.c]: you missed:
rte_free(relay->pr_netdev->name)

File: lib/netdev-rte-offloads.c [hunk 19]
```diff
@@ -1778,3 +1788,838 @@ netdev_rte_offload_preprocess(struct dp_packet *packet, uint32_t mark)
...
+    action_rss.queue_num = j;
+
+    action[0].type = RTE_FLOW_ACTION_TYPE_RSS;
+    action[0].conf = &action_rss;
+    action[1].type = RTE_FLOW_ACTION_TYPE_END;
+
+    pattern[0].type = RTE_FLOW_ITEM_TYPE_ETH;
+    pattern[0].spec = NULL;
+    pattern[0].mask = NULL;
+    pattern[1].type = RTE_FLOW_ITEM_TYPE_END;
```
Comment (elibritstein) [inline lib/netdev-rte-offloads.c]: can init those at the definitions of them

File: lib/netdev-rte-offloads.c [hunk 20]
```diff
@@ -1778,3 +1788,838 @@ netdev_rte_offload_preprocess(struct dp_packet *packet, uint32_t mark)
...
+            if (rte_atomic16_test_and_set(&relays[relay_id].valid)) {
+                next_valid = relay_id;
+            }
+        }
+        /* check that the socket/PCI doesn't exist */
+        if (netdev_rte_offloads_check_valid_params(pci, vhost_socket, relay_id) < 0) {
+            if (next_valid != NETDEV_RTE_OFFLOADS_INVALID_ID) {
+                rte_atomic16_clear(&relays[next_valid].valid);
+            }
+            return -1;
```
Comment (elibritstein) [inline lib/netdev-rte-offloads.c]: if you have NETDEV_RTE_OFFLOADS_INVALID_ID, why not return it here?

File: lib/netdev-rte-offloads.c [hunk 21]
```diff
@@ -1778,3 +1788,838 @@ netdev_rte_offload_preprocess(struct dp_packet *packet, uint32_t mark)
...
+    if (ret) {
+        VLOG_ERR("port_init failed, port_id %u\n", *port_id_vm);
+        rte_eth_dev_stop(*port_id_vf);
+        rte_eth_dev_close(*port_id_vf);
+        return ret;
+    }
+    return 0;
+}
+
+static int
```
Comment (elibritstein) [inline lib/netdev-rte-offloads.c]: it can be void function in this commit

File: lib/netdev-rte-offloads.c [hunk 22]
```diff
@@ -2206,7 +2206,8 @@ netdev_rte_offloads_update_relay(uint16_t relay_id,
 
 static int
 netdev_rte_offloads_forward_traffic(struct netdev_rte_offloads_qpair *qpair,
-                                    uint16_t queue_id)
+                                    uint16_t queue_id,
+                                    uint32_t *fwd_rx)
```
Comment (elibritstein) [inline lib/netdev-rte-offloads.c]: can use returned value for this function as well

File: lib/netdev-rte-offloads.c (hunk 15 above)
Comment (noaezra) [reply lib/netdev-rte-offloads.c]: all of them are needed.
I moved them to bottom.

File: lib/netdev-dpdk.c (hunk 14 above)
Comment (noaezra) [reply lib/netdev-dpdk.c]: it's not an error, it reset the function to be null.

File: lib/netdev-dpdk.c (hunk 16 above)
Comment (noaezra) [reply lib/netdev-dpdk.c]: I don't understand what you mean here

Comment (elibritstein) [reply lib/netdev-dpdk.c]: it's a coding convention, though i see it's not in OVS one.
it means that order of declaration of variable should be longest to shortest
